cursor.go

```diff
@@ -0,0 +1,173 @@
+package promptui
+
+import "fmt"
+
+// Pointer is A specific type that translates a given set of runes into a given
+// set of runes pointed at by the cursor.
+type Pointer func(to []rune) []rune
+
+func defaultCursor(ignored []rune) []rune {
+	return []rune("\u2588")
+}
+
+func blockCursor(input []rune) []rune {
+	return []rune(fmt.Sprintf("\\e[7m%s\\e[0m", string(input)))
+}
+
+func pipeCursor(input []rune) []rune {
+	marker := []rune("|")
+	out := []rune{}
+	out = append(out, marker...)
+	out = append(out, input...)
+	return out
+}
+
+var (
+	// DefaultCursor is a big square block character. Obscures whatever was
+	// input.
+	DefaultCursor Pointer = defaultCursor
+	// BlockCursor is a cursor which highlights a character by inverting colors
+	// on it.
+	BlockCursor Pointer = blockCursor
+	// PipeCursor is a pipe character "|" which appears before the input
+	// character.
+	PipeCursor Pointer = pipeCursor
+)
+
+// Cursor tracks the state associated with the movable cursor
+// The strategy is to keep the prompt, input pristine except for requested
+// modifications. The insertion of the cursor happens during a `format` call
+// and we read in new input via an `Update` call
+type Cursor struct {
+	// shows where the user inserts/updates text
+	Cursor Pointer
+	// what the user entered, and what we will echo back to them, after
+	// insertion of the cursor and prefixing with the prompt
+	Input []rune
+	// Put the cursor before this slice
+	Position int
+}
+
+// NewCursor create a new cursor, with the DefaultCurso, the specified input,
+// and position at the end of the specified starting input.
+func NewCursor(startingInput string, pointer Pointer) Cursor {
+	if pointer == nil {
+		pointer = defaultCursor
+	}
+	return Cursor{Cursor: pointer, Position: len(startingInput), Input: []rune(startingInput)}
+}
+
+func (c *Cursor) String() string {
+	return fmt.Sprintf(
+		"Cursor: %s, Input %s, Position %d",
+		string(c.Cursor([]rune(""))), string(c.Input), c.Position)
+}
+
+// End is a convenience for c.Place(len(c.Input)) so you don't have to know how I
+// indexed.
+func (c *Cursor) End() {
+	c.Place(len(c.Input))
+}
+
+// Start is convenience for c.Place(0) so you don't have to know how I
+// indexed.
+func (c *Cursor) Start() {
+	c.Place(0)
+}
+
+// ensures we are in bounds.
+func (c *Cursor) correctPosition() {
+	if c.Position > len(c.Input) {
+		c.Position = len(c.Input)
+	}
+
+	if c.Position < 0 {
+		c.Position = 0
+	}
+}
+
+// insert the cursor rune array into r before the provided index
+func format(a []rune, c *Cursor) string {
+	i := c.Position
+	var b []rune
+
+	out := make([]rune, 0)
+	if i < len(a) {
+		b = c.Cursor([]rune(a[i : i+1]))
+		out = append(out, a[:i]...)   // does not include i
+		out = append(out, b...)       // add the cursor
```

Comment:
I think we don't necessarily always want to include the cursor in the output -- we want it for display on the screen, but we don't want e.g. a prompt value to also contain the cursor rune. This breaks for example the `_examples/prompt/main.go`, because it expects to be able to format a number from the prompt text, but the cursor rune causes it to fail.

Reply:
Have another look, I think it should work now.

Did anyone attempt to implement the ScreenBuffer as a readline Painter? It seems like we could delete most of the logic I'm working if we used more of the readline functionality.

Reply:
> Did anyone attempt to implement the ScreenBuffer as a readline Painter? It seems like we could delete most of the logic I'm working if we used more of the readline functionality.

I'm not honestly sure -- it's possible sometime in the past, but no one has tried recently. We'd be more than happy to take a look at a PR for it though ;)